File: template-method/dart/compiler-arm.dart
import "compiler.dart";

class CompilerARM extends Compiler {

    CompilerARM() {
		print( "\n\nInitialize ARM Compiler\n");
	}

    @override
	String preprocess(String code) {
		String output = "void main() { cout << 32767 << endl; }";
		print( "ARM preprocesor generated '$output'");
		return output;
	}

    @override
	String compile(String code) {
		String output = ".asciz \"32767\\n\"";
		print( "ARM compiler generated '$output'");
		return output;
	}

    @override
	String assemble(String code) {
        String output = _toBinary(code);
		print( "ARM assembler generated '$output'");
		return output;
	}

    @override
	String link(String code) {
		String output = _toBinary("CompilerARM")+code;
		print( "ARM linker generated '$output'");
		return output;
	}

    String _toBinary(String code) {
        return code.codeUnits.map((int unit) => unit.toRadixString(2)).toList().join().substring(0,12);
    }
}

File: template-method/python3/compiler_arm.py
from compiler import *

class CompilerARM(Compiler):
	def __init__(self):
		super()
		print("\n\nInitialize ARM Compiler\n\n")

	def preprocess(self, code):
		output = 'void main() { cout << 32767 << endl; }'
		print("ARM preprocesor generated '{}'".format(output))
		return output

	def compile(self, code):
		output = '.asciz "32767\\n"'
		print("ARM compiler generated '{}'".format(output))
		return output

	def assemble(self, code):
		output = self._to_binary(code)
		print("ARM assembler generated {}".format(output))
		return output

	def link(self, code):
		output = self._to_binary("CompilerARM")+code
		print("ARM linker generated {}".format(output))
		return output

	def _to_binary(self, code):
		code_bytes = ''
		for char in code:
			code_bytes += format(ord(char), '08b')
		return code_bytes[:12]
